Gather-logs enhancements

donnyyung opened the pull request:
## Description
Currently a WIP, only added the PR so I could test some circle changes


This adds a few features to gather-logs:
1. The ability to anonymize logs pod names + namespaces from the logs you submit (we'll work on anonymizing even more things in the future)
2. Adding a flag to gather-logs that will get the pod yaml for the traffic-manager + pods that contain a traffic-agent... this can be very helpful in debugging

## Checklist

<!--
  Please review the requirements for each checkbox, and check them
  off (change "[ ]" to "[x]") as you verify that they are complete.
-->

 - [x] I made sure to update `./CHANGELOG.md`.
 - [ ] I made sure to either submit a docs PR, or tell Matt about the necessary documentation changes.
 - [ ] My change is adequately tested.
 - [x] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently.
 - [x] I updated `TELEMETRY.md` if I added, changed, or removed a metric name.

Commit:
Add anonymize flag to gather-logs

I will add the anonymization features in multiple commits. This first
one enables anonymizing the pod names when writing them to the zip file
that is output

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Support getting pod yaml in gather-logs

This adds the ability for users to opt-in to sending the pod yaml for
the traffic-manager and pods containing a traffic-agent.

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Add changelog

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Making functions less complex so lint is happy

Also updated some comments to make things clearer

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Link directly to github issue creation

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Update tests for pod yaml in gather-logs

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Unit tests for anonymize logs

Signed-off-by: Donny Yung <donaldyung@datawire.io>

josecv commented on pkg/client/cli/cmd_gatherlogs.go:
```diff
@@ -119,6 +139,7 @@ func (gl *gatherLogsArgs) gatherLogs(ctx context.Context, cmd *cobra.Command, st
 	scout.SetMetadatum("daemon_logs", daemonLogs)
 	scout.SetMetadatum("traffic_manager_logs", gl.trafficManager)
 	scout.SetMetadatum("traffic_agent_logs", gl.trafficAgents)
+	scout.SetMetadatum("get_pod_yaml", gl.podYaml)
```
let's add a metadatum for the anonymize flag too, just for completeness' sake

josecv commented on pkg/client/cli/cmd_gatherlogs.go:
```diff
@@ -203,6 +226,40 @@ func (gl *gatherLogsArgs) gatherLogs(ctx context.Context, cmd *cobra.Command, st
+		}
+		defer fd.Close()
+		fdWriter := bufio.NewWriter(fd)
+		_, _ = fdWriter.WriteString(content)
```
should we be throwing out this error? might make sense to at least report it in some capacity (unless it can't happen?)

donnyyung replied:
Oh whoops that was my bad. I error checked in the other place I used this, not sure why I didn't here. will fix

josecv commented on pkg/client/cli/cmd_gatherlogs.go:
```diff
@@ -260,12 +317,122 @@ func zipFiles(files []string, zipFileName string) error {
+	// so we can anonymize the namespace
+	nameComponents := strings.SplitN(podName, ".", 2)
+	if len(nameComponents) != 2 {
+		unknownPodName := "anonPod.anonNamespace"
```
hmm could we give this an ordinal based on `len(anonymizer.podNames)` too? im worried that if for whatever  reason this happens to one pod, it will happen to  all of em and we won't be able to distinguish them

donnyyung replied:
Ah great catch. Will do.

josecv commented on pkg/client/cli/cmd_gatherlogs.go:
```diff
@@ -260,12 +317,122 @@ func zipFiles(files []string, zipFileName string) error {
+
+		// Strip the hash off of the pod name + replace remainder with anonymized name
+		brokenPodName := strings.Split(podName, "-")
+		podName = strings.Join(brokenPodName[:len(brokenPodName)-2], "-")
```
watch out cause this is not going to work for statefulsets -- if a statefulset is called `my-sts` the pods will be `my-sts-0`, `my-sts-1`, etc -- you might need a regex of some sort to check what kind of pod it is

this will similarly have trouble for pods created from Jobs which also have their own format IIRC

donnyyung replied:
ah yeah let me think about that... I don't think we need to worry about Jobs (yet) since we can't intercept those (afaik), but it won't hurt to be forward thinking if it's not too difficult.

josecv commented on pkg/client/cli/cmd_gatherlogs.go:
```diff
@@ -260,12 +317,122 @@ func zipFiles(files []string, zipFileName string) error {
+		// Strip the namespace off of the podName
+		podParts := strings.Split(fullPodName, ".")
+
+		// replace pod name + full hash with the anonymized name
+		podName := podParts[0]
+		stringContent = strings.ReplaceAll(stringContent, podName, anonPodName)
+
+		// Strip the hash off of the pod name + replace remainder with anonymized name
```
wait im confused why do we do both of these things? it seems either/or would be sufficient?

donnyyung replied:
Ah I should better describe this in the logs.. So the problem was in the traffic-manager logs we might have something like this:
```
2021-10-01 18:45:36.8322 info    agent-injector : Injecting traffic-agent into pod echo-auto-inject-869ff95bc7-      .default
```
but in the traffic-agent we have logs like this:
```
2021-10-01 18:45:38.3415 info    {Name:echo-auto-inject Namespace:default PodIP:10.42.0.35 AgentPort:9900 AppMo      unts:/tel_app_mounts AppPort:8080 ManagerHost:traffic-manager.ambassador ManagerPort:8081}
```

So I could just ReplaceAll without the hash, but then the first log would look like:
```
2021-10-01 18:45:36.8322 info    agent-injector : Injecting traffic-agent into pod pod-1-869ff95bc7-.namespace-2:
```

Although I just noticed my code wouldn't anonymize this bc it doesn't include the 2nd part of the hash...

Hm I think I have an idea that might remedy this + the issue with statefulsets

josecv reviewed (changes requested):
v nice work overall just some concerns -- also i think this is missing the timestamp change (unless im dumb and i missed it which is always possible) so i can re-review when that is added 🙂

donnyyung commented:
I had gotten called into the other flurry of alerts before I could add the timestamp fixes :( will add that first and then go through your comments, thanks Jose!

Commit:
gather-logs shouldn't fail if connector does

Instead we let users know that we were not able to get the logs from
components in the kubernetes cluster and return a zip file containing
all the logs we were able to get.

Signed-off-by: Donny Yung <donaldyung@datawire.io>

Commit:
Fix file info in zip

Zipfile now has correct data about when the files in the zip were
created (e.g. not 1979 which is the default value they were using
before)

Signed-off-by: Donny Yung <donaldyung@datawire.io>

donnyyung commented:
Fixes https://github.com/telepresenceio/telepresence/issues/2126

Commit:
PR feedback

Signed-off-by: Donny Yung <donaldyung@datawire.io>

josecv commented on pkg/client/cli/cmd_gatherlogs.go:
```diff
@@ -239,7 +242,10 @@ func writeResponseToFiles(lr *manager.LogsResponse, anonymizer *anonymizer, expo
-		_, _ = fdWriter.WriteString(content)
+		_, err = fdWriter.WriteString(content)
+		if err != nil {
+			return nil
```
should this return err?

donnyyung replied:
🤦‍♂️

josecv commented on k8s/echo-easy.yaml:
```diff
@@ -2,7 +2,7 @@
 kind: Service
 metadata:
-  name: echo-easy
+  name: "a123f"
```
hmm is this an accident? it looks a little like you were testing the `getSignificantPodNames` locally; it would probably be a good idea to actually add this to the unit tests

josecv commented on pkg/client/cli/cmd_gatherlogs.go:
```diff
@@ -455,3 +463,43 @@ func anonymizeLog(stdout io.Writer, logFile string, anonymizer *anonymizer) erro
+	subNames := make([]string, 2)
+	nameComponents := strings.Split(podName, "-")
+	significantComponent := false
+	for i, component := range nameComponents {
+		// Indicates we found a Pod that comes from either a
+		// ReplicaSet and/or a Deployment
+		if _, err := hex.DecodeString(component); err == nil {
+			significantComponent = true
+		}
+		// Indicates we found a Pod that comes from a StatefulSet
+		if _, err := strconv.Atoi(component); err == nil {
+			significantComponent = true
+		}
+
+		if significantComponent {
+			nameComponents = nameComponents[:i+1]
+			break
+		}
+	}
```
hmm im not sure im sold -- what if i have a deployment called `hello-0`? i almost wanna say we should loop in reverse; what about something like this:

```go
lastSignificant := false
subNames := []string{}
for i := len(nameComponents)-1; i >= 0; i-- {
                component := nameComponents[i]
                significant := false
                // Indicates we found a Pod that comes from either a
                // ReplicaSet and/or a Deployment
                if _, err := hex.DecodeString(component); err == nil {
                    significant = true
                }
                // Indicates we found a Pod that comes from a StatefulSet
                if _, err := strconv.Atoi(component); err == nil && !lastSignificant {
                    significant = true
                }
                
                if !significant && lastSignificant {
                     // my-sts or my-deploy
                     subNames = append(subNames, strings.Join(nameComponents[:i], "-"))
                     break
                } else if significant {
                     // my-sts-0 or my-deploy-asdf-asdf or my-deploy-asdf
                     subNames = append(subNames, strings.Join(nameComponents[:i], "-"))
                     lastSignificant = true
                }
}
```

although now that i look at that i think it won't work for a deployment called `a123f` 🤔 it might be easier to just do this with regexes 😅 should be easy to make one for statefulsets, one for deployments, and one for replicasets and see which is matched

donnyyung replied:
yeah there wasn't really any way that I was super happy with. I will try to think of a regexes.

Commit:
PR feedback v2

Mostly fixing up the getSignificantPodNames to use regex instead of
doing a (hacky) string parsing

Signed-off-by: Donny Yung <donaldyung@datawire.io>

donnyyung commented on pkg/client/cli/cmd_gatherlogs.go:
```diff
@@ -469,37 +468,42 @@ func anonymizeLog(stdout io.Writer, logFile string, anonymizer *anonymizer) erro
+	replicaSetRegex := regexp.MustCompile("(?P<appName>.*)-(?P<hash>[0-9a-f]+)-([0-9a-z]{5})$")
+	sigNames := []string{}
+	switch {
+	case statefulSetRegex.MatchString(podName):
```
this seems ridiculously complicated to get matches but I couldn't find another way to do it?? If another way exists, I'd love to know.

josecv commented on pkg/client/cli/cmd_gatherlogs.go:
```diff
@@ -469,37 +468,29 @@ func anonymizeLog(stdout io.Writer, logFile string, anonymizer *anonymizer) erro
+		appName := match[1]
+		rsName := fmt.Sprintf("%s-%s", appName, match[2])
+		// add the app name with and without generated ReplicaSet hash
+		sigNames = append(sigNames, rsName, appName)
```
should we also append podName here?

donnyyung replied:
I hadn't seen the full pod name show up in any logs, but it can't hurt to be extra thorough, will do :)

josecv reviewed (approved):
lgtm with just one minor issue

Commit:
include podName in getSignificantPodNames

Signed-off-by: Donny Yung <donaldyung@datawire.io>
